### ISSUE TITLE
group/user add --root option when target arch is different than host

### ISSUE THREAD

**Issue by fariouche** [shadow-maint/shadow#18]
Hi,

It is a long time I'm missing this feature: being able to create user/group when cross compiling.
Currently the --root option is doing a chroot.
I would like to add a new option (---prefix?) that will just prefix the path to shadow files instead of using the hardcoded /etc/group etc paths.
In this case we'll be able to create all the users/group before generating the embedded firmware.
Of course, it may imply some limitations (pam authentification will still use the host accounts for example... same for ldap?)
But before going this route, I would like to understand if there is any reason why was it never done?
Should I take core of some specific points too? (maybe some advanced features that I should be aware of?)

Thank you

**Comment by hallyn** [comment]
Hi,

just a thought - you could actually mount an overlayfs with read-only /
under an empty rw directory, and pass that to --chroot, right?  Then
umount the overlay and end up with just the changed files which you
could copy into the target rootfs later?

I can't speak as to whether there was a historic reason to not do what
you suggest, hopefully @bubulle or @nekral can.

**Comment by fariouche** [comment]
Hi,
hum. That means copy the target's passwd/group/shadow etc file to the temporary overlay and then chroot to it. That may work actually. My only concern is the sandboxing mechanism of the packaging system (this is for gentoo): it may not be allowed. I will ask them.
By the way, I tried a quick proof of concept by adding --prefix, and it works (no NIS nor LDAP of course)
I will check for the simplest route

**Comment by nekral** [comment]
Hi,

The intent of the --root option was to use the config files, and (e.g. PAM, NSS) libraries from the chroot.

I'm not sure the PAM config files / libraries would be relevant for useradd, but I would be careful on the behavior expected regarding
* files on the host. Make sure to overlay all possible user / group databases (in addition to group/passwd/shadow, you will need gshadow/subuid/subgid)
* /etc/login.defs (e.g. range of valid user/group IDs)
* other databases used by the target or host (worst case, you create a user or group with an ID already in use on the target)

Those are probably less critical:
* log files / audit reports
* backup files
* lock files

I don't think I though about everything. I would recommend to check all open files / libraries (e.g. strace) during the execution to make sure it behaves as expected.

**Comment by fariouche** [comment]
Thank you, I see what you mean.

Just for information, this is for gentoo, when creating a target (x86 or cross compiled) and maybe usable for any embedded firmware creation tools... In that case, the audit and logging are maybe not relevant here...
For NSS/LDAP, it also does not make sense, I think, to use the target configuration files (as the user/group creation will be performed by the host)

Making gentoo's build system depends on recent kernel overlayfs is unfortunately not possible. Plus the fact that the sandboxing security will prevent this usage.

Maybe I will end up with a specfic re-write of the user/group creation in gentoo directly, but this will not benefit others....

**Comment by fariouche** [comment]
My proposal is as follows:
1- if --prefix is not specified, the behavior will not be changed (no possible regression)
2- groupadd, useradd  will be modified to create a user and group and verify the user/group existence inside the --prefix folder
3- For NIS/LDAP, they can be checked using the host configuration files.
4- For PAM authentication, the host pam module will be used.
5- the --prefix folder may not be based on shadow utilities. (busybox for example, no login.defs etc)
6- --prefix may be of any architecture (so we cannot use the --root option of groupadd to chroot)
7- no foreign dependency like a mandatory kernel option (overlayfs) or any external library will be added.
8- lock and backup files will be created inside --prefix folder. All other files will be in the host (log,audit, login.defs etc)

We cannot use the getent function as is (part of the glibc), as this one uses hardcoded path to the databases.
I can use fgetgrent and fgetpwdent functions, + some other functions to add NIS/LDAP checks in replacement of the gentent function ...

If you feel like any of the points above is not right, or that this --prefix feature finally does not make sense to the shadow utility, please let me know.

**Comment by fariouche** [comment]
Hi guys,

It was bigger than expected, but I think I have something working.
I'm currently testing it.

For the implementation:
I added --prefix option.
When this option is set (and it cannot be set to "/"), it will prefix all path for the following files:
shadow/passwd,group,sgroup,login.defs and the sub uid/gid file.

Now, for the behavior, it is easy.
- SELINUX is ignored for now (if you try to specify -Z option, it will complain and exit). The reason is that it seems that selinux cannot be configured to output to some directory and also I cannot test.
- PAM: the host configuration will be use (PAM is used to check the user credentials, so it does not make sense here to use pam configuration from the prefix directory)
- NIS: it may be possible to use the prefix configuration files, but I cannot implement/modify something that I cannot test. So for now, if the prefix option is used, I explicitly disable NIS checking.
- LDAP: totally ignored (in fact I didn't see any LDAP stuff in the code...)
- LOG/audit: using host files. I believe that is does not make sense to log in the prefix directory.

Fortunately the internal code in shadow was already ready to specify custom database path.
So I only had to had wrapper functions for the different needed functions, and override the databases name.
If prefix is not specified, the exact same old functions are used, so that no possible regression is possible. If a prefix is specified, some functions (get*nam(), get*uid() etc) are redirected to internal one, defined in libmisc/prefix_flag.c

If later we had NIS or SELINUX support, it will be only a matter of a few lines in prefix_flag.c

As soon as I finish the tests, I will submit a pullrequest.
